# Scraping Yahoo Finance

## Week 7. Practice Programming Assignment 1

In this assignment you are required to look at historical data for 30 companies from [Dow Jones Index](https://en.wikipedia.org/wiki/Dow_Jones_Industrial_Average). Tickers for companies from the index can be found in *dow_jones_tickers.txt*. For each company you should get historical daily stock prices for 2019 from https://finance.yahoo.com/, and then use the data to answer the questions you will find below. 

### Coding part

In [1]:
# you can start your web-scraping here
import re
import requests
import pandas as pd
import numpy as np
import time
from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.chrome.service import Service as ChromeService
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.select import Select
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.chrome.service import Service
from webdriver_manager.chrome import ChromeDriverManager

In [2]:
dow_jones_companies = []

with open('dow_jones_tickers.txt') as f:
    for ticker in f:
        dow_jones_companies.append(ticker.strip())

print("Dow Jones Companies:\n", *dow_jones_companies)
print("Number of Companies:", len(dow_jones_companies))

Dow Jones Companies:
 AXP AAPL BA CAT CSCO CVX XOM GS HD IBM INTC JNJ KO JPM MCD MMM MRK MSFT NKE PFE PG TRV UNH RTX VZ V WBA WMT DIS DOW
Number of Companies: 30


In [3]:
def get_stock_info(company):
    
    # url = 'https://finance.yahoo.com/quote/' + ticker + '/history?p=' 
    url = 'https://finance.yahoo.com/quote/' + company + '/history?period1=1546300800&period2=1577836800&interval=1d&filter=history&frequency=1d&includeAdjustedClose=true'

    driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()))
    
    driver.get(url)
    driver.implicitly_wait(5)
    
    # Click on 'I agree' button and wait for a second.
    driver.find_element('xpath', '//button[@value="agree"]').click()
    time.sleep(3)

    i = 0
    while i < 4:
        
        driver.find_element(By.TAG_NAME, 'html').send_keys([Keys.PAGE_DOWN for i in range(100)])
        time.sleep(3)
        i += 1
     
    
    soup = BeautifulSoup(driver.page_source)
    results = soup.find_all('tr', {'class': 'BdT Bdc($seperatorColor) Ta(end) Fz(s) Whs(nw)'})   
    
    for row in results:
        record = row.find_all('span')
        if len(record) != 7: continue
        
        data.append({'Ticker': ticker,
                     'Date': record[0].text,
                     'Open': float(record[1].text),
                     'High': float(record[2].text),
                     'Low': float(record[3].text),
                     'Close': float(record[4].text),
                     'Adj Close': float(record[5].text),
                     'Volume': int(record[6].text.replace(',', ''))})
        
     
    driver.quit()

In [5]:
#data = []
#get_stock_info("DOW")
#data_df = pd.DataFrame(data)
#data_df['Date'] = pd.to_datetime(data_df['Date'])
#data_df['Month'] = data_df['Date'].dt.month
#data_df.shape

In [10]:
data = []

for ticker in dow_jones_companies:
    print('', ticker, end = '')
    get_stock_info(ticker)
    time.sleep(2)
    
data_df = pd.DataFrame(data)
data_df['Date'] = pd.to_datetime(data_df['Date'])
data_df['Month'] = data_df['Date'].dt.month

 AXP


[WDM] - ====== WebDriver manager ======


2022-07-12 01:58:45,802 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 01:58:45,844 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 01:58:45,846 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 01:58:45,911 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 AAPL


[WDM] - ====== WebDriver manager ======


2022-07-12 01:59:12,764 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 01:59:12,801 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 01:59:12,802 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 01:59:12,864 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 BA


[WDM] - ====== WebDriver manager ======


2022-07-12 01:59:39,239 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 01:59:39,277 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 01:59:39,279 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 01:59:39,342 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 CAT


[WDM] - ====== WebDriver manager ======


2022-07-12 02:00:06,164 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:00:06,204 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:00:06,205 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:00:06,270 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 CSCO


[WDM] - ====== WebDriver manager ======


2022-07-12 02:00:31,995 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:00:32,033 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:00:32,034 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:00:32,095 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 CVX


[WDM] - ====== WebDriver manager ======


2022-07-12 02:00:59,936 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:00:59,973 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:00:59,974 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:01:00,038 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 XOM


[WDM] - ====== WebDriver manager ======


2022-07-12 02:01:26,293 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:01:26,333 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:01:26,334 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:01:26,396 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 GS


[WDM] - ====== WebDriver manager ======


2022-07-12 02:01:53,501 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:01:53,536 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:01:53,538 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:01:53,605 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 HD


[WDM] - ====== WebDriver manager ======


2022-07-12 02:02:20,844 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:02:20,878 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:02:20,879 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:02:20,940 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 IBM


[WDM] - ====== WebDriver manager ======


2022-07-12 02:02:47,236 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:02:47,273 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:02:47,274 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:02:47,343 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 INTC


[WDM] - ====== WebDriver manager ======


2022-07-12 02:03:14,406 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:03:14,445 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:03:14,446 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:03:14,511 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 JNJ


[WDM] - ====== WebDriver manager ======


2022-07-12 02:03:40,002 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:03:40,039 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:03:40,040 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:03:40,103 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 KO


[WDM] - ====== WebDriver manager ======


2022-07-12 02:04:06,447 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:04:06,485 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:04:06,486 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:04:06,545 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 JPM


[WDM] - ====== WebDriver manager ======


2022-07-12 02:04:39,479 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:04:39,520 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:04:39,522 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:04:39,585 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 MCD


[WDM] - ====== WebDriver manager ======


2022-07-12 02:05:05,423 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:05:05,463 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:05:05,465 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:05:05,526 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 MMM


[WDM] - ====== WebDriver manager ======


2022-07-12 02:05:31,447 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:05:31,484 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:05:31,485 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:05:31,550 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 MRK


[WDM] - ====== WebDriver manager ======


2022-07-12 02:06:08,377 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:06:08,415 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:06:08,416 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:06:08,481 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 MSFT


[WDM] - ====== WebDriver manager ======


2022-07-12 02:06:39,911 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:06:39,947 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:06:39,949 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:06:40,009 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 NKE


[WDM] - ====== WebDriver manager ======


2022-07-12 02:07:06,159 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:07:06,197 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:07:06,198 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:07:06,261 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 PFE


[WDM] - ====== WebDriver manager ======


2022-07-12 02:07:32,991 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:07:33,028 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:07:33,029 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:07:33,087 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 PG


[WDM] - ====== WebDriver manager ======


2022-07-12 02:07:59,416 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:07:59,454 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:07:59,456 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:07:59,516 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 TRV


[WDM] - ====== WebDriver manager ======


2022-07-12 02:08:26,932 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:08:26,968 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:08:26,969 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:08:27,026 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 UNH


[WDM] - ====== WebDriver manager ======


2022-07-12 02:08:53,823 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:08:53,863 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:08:53,864 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:08:53,919 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 RTX


[WDM] - ====== WebDriver manager ======


2022-07-12 02:09:21,026 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:09:21,065 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:09:21,066 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:09:21,124 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 VZ


[WDM] - ====== WebDriver manager ======


2022-07-12 02:09:47,529 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:09:47,565 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:09:47,567 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:09:47,629 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 V


[WDM] - ====== WebDriver manager ======


2022-07-12 02:10:14,539 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:10:14,578 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:10:14,579 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:10:14,639 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 WBA


[WDM] - ====== WebDriver manager ======


2022-07-12 02:10:40,902 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:10:40,939 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:10:40,940 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:10:41,008 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 WMT


[WDM] - ====== WebDriver manager ======


2022-07-12 02:11:08,736 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:11:08,773 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:11:08,774 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:11:08,835 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 DIS


[WDM] - ====== WebDriver manager ======


2022-07-12 02:11:34,758 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:11:34,796 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:11:34,797 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:11:34,894 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache
 DOW


[WDM] - ====== WebDriver manager ======


2022-07-12 02:12:01,895 INFO ====== WebDriver manager ======


[WDM] - Current google-chrome version is 103.0.5060


2022-07-12 02:12:01,935 INFO Current google-chrome version is 103.0.5060


[WDM] - Get LATEST chromedriver version for 103.0.5060 google-chrome


2022-07-12 02:12:01,936 INFO Get LATEST chromedriver version for 103.0.5060 google-chrome


[WDM] - Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


2022-07-12 02:12:02,000 INFO Driver [/home/nata/.wdm/drivers/chromedriver/linux64/103.0.5060.53/chromedriver] found in cache


In [160]:
data_df

Ticker       Date    Open    High     Low   Close  Adj Close   Volume  \
0       AXP 2019-12-31  124.29  124.57  123.78  124.49     119.80  2340400   
1       AXP 2019-12-30  125.20  125.46  124.18  124.30     119.62  2306500   
2       AXP 2019-12-27  125.84  125.97  125.11  125.19     120.48  1788600   
3       AXP 2019-12-26  124.98  125.44  124.53  125.41     120.69  1486600   
4       AXP 2019-12-24  124.95  125.33  124.38  124.74     120.05   953500   
...     ...        ...     ...     ...     ...     ...        ...      ...   
7502    DOW 2019-03-26   49.00   49.75   48.18   48.85      41.12   504700   
7503    DOW 2019-03-25   48.60   49.40   48.00   49.15      41.37   440900   
7504    DOW 2019-03-22   48.80   49.95   48.16   48.60      40.91   844700   
7505    DOW 2019-03-21   49.99   50.00   48.20   48.98      41.23  1764700   
7506    DOW 2019-03-20   52.75   53.50   49.50   49.80      41.92  2350800   

      Month  year   average  
0        12  2019  0.160914  
1        12  2019 -0.718850  
2        12  2019 -0.516529  
3        12  2019  0.344055  
4        12  2019 -0.168067  
...     ...   ...       ...  
7502      3  2019 -0.306122  
7503      3  2019  1.131687  
7504      3  2019 -0.409836  
7505      3  2019 -2.020404  
7506      3  2019 -5.592417  

[7507 rows x 11 columns]

In [161]:
new_df = data_df

new_df['year'] = pd.DatetimeIndex(new_df['Date']).year
new_df.shape

(7507, 11)

In [162]:
df_2019 = new_df.drop(new_df[new_df['year'] != 2019].index)
df_2019.shape

(7507, 11)

In [163]:
df_2019_sort = df_2019.sort_values(by=['Ticker', 'Date'] , ascending=True, na_position='first')
# df_2019_sort[0:252]

In [164]:
df_df = df_2019_sort.groupby(['Ticker'])['Ticker'].count()
df_df.values

array([252, 252, 252, 252, 252, 252, 252, 199, 252, 252, 252, 252, 252,
       252, 252, 252, 252, 252, 252, 252, 252, 252, 252, 252, 252, 252,
       252, 252, 252, 252])

In [172]:
dub = df_2019_sort.duplicated()
dub2 = list(dub.values)
check = []
for el in dub2:
    if el not in check:
        check.append(el)
check

[False]

In [168]:
df_2019_sort[df_2019_sort['Ticker'] == 'AAPL']['Close'].iloc[-1]

73.41

In [169]:
df_2019_sort[df_2019_sort['Ticker'] == 'AAPL']['Open'].iloc[0]

38.72

In [18]:
df_2019_sort.shape

(7507, 10)

In [20]:
average_change = []
for ticker in dow_jones_companies:
    average_change.append((df_2019_sort[df_2019_sort['Ticker'] == ticker]['Close'].iloc[-1] / df_2019_sort[df_2019_sort['Ticker'] == ticker]['Open'].iloc[0] - 1) * 100)

# average_change

<br><br>

### Questions

<br><br>

**Question 1.** What is the average change of price over the year (in %)?

*Note 1*. The opening price is the price at which a stock first trades upon the opening of an exchange on a trading day.

*Note 2*. The closing price for any stock is the final price at which it trades during regular market hours on any given day.

*Note 3*. Here by the price change we going to mean a ratio of a closing price in the last day of the period to an opening price in the first day of that period, subtracted one and multiplied by 100.

Example: if a price of a stock in day 1 opened at \\$100, and its close price in the last day was \\$120, then the price change during the period is: $$\left(\dfrac{120}{100}-1\right) * 100 = (1.2 - 1) * 100=20.$$

The price grew by 20%.

In [21]:
answer_part_1 = 23.66501620232888

In [23]:
answer_part_1 = np.mean(average_change)
answer_part_1
# 23.66501620232888

23.66501620232888

<br>

**Question 2.** What company's stock price grew the most (in %)? Enter ticker of the company as an answer)

In [26]:
answer_part_2 = 'AAPL'

In [27]:
# Question 2. What company's stock price grew the most (in %)?
#part_2 = dow_jones_companies[average_change.index(max(average_change))]
part_2 = dow_jones_companies[np.argmax(average_change)]
part_2

'AAPL'

<br>

**Question 3.** What company's stock lost in price the most (in %)? Enter ticker of the company as an answer

In [30]:
answer_part_3 = 'WBA'

In [31]:
# Question 3. What company's stock lost in price the most (in %)?
answer_part_3 = dow_jones_companies[np.argmin(average_change)]
answer_part_3

'WBA'

<br>

**Question 4.** What company had the largest summary volume over the year? Enter ticker of the company as an answer

In [34]:
answer_part_4 = 'AAPL'

In [35]:
# Question 4. What company had the largest summary volume over the year? 
part_4 = data_df.groupby('Ticker')['Volume'].sum().sort_values(ascending=False).reset_index()
# part_4
answer_part_4 = part_4.loc[0, 'Ticker']
answer_part_4

'AAPL'

<br>

**Question 5.** What is the biggest stock price daily increase (in %)? Enter the number 

In [39]:
answer_part_5 = 7.740997118000379

In [40]:
df_2019_sort['daily increase'] = (df_2019_sort['Close'] - df_2019_sort['Open']) / df_2019_sort['Open']  * 100
#df_2019_sort

In [41]:
df_2019_sort['daily increase'].max()

7.740997118000379

<br><br>

**Question 6.** What is the company that had the biggest stock price daily increase? Enter ticker of the company as an answer

answer_part_6 = 'BA'

In [42]:
df_2019_sort[df_2019_sort['daily increase'] == 7.740997118000379]

Ticker       Date    Open    High     Low   Close  Adj Close    Volume  \
709     BA 2019-03-11  371.27  402.67  365.55  400.01     390.64  34742200   

     Month  year  daily increase  
709      3  2019        7.740997

<br>

**Question 7.** What is the biggest stock price daily decrease (in %)? Enter the number

In [43]:
answer_part_7 = -7.163564556111081

In [44]:
df_2019_sort['daily increase'].min()

-7.163564556111081

In [45]:
df_2019_sort[df_2019_sort['daily increase'] == -7.163564556111081]

Ticker       Date    Open    High     Low   Close  Adj Close    Volume  \
5723    UNH 2019-04-16  238.01  238.01  215.82  220.96     210.23  27361400   

      Month  year  daily increase  
5723      4  2019       -7.163565

<br>

**Question 8.** What is the company that had the biggest stock price daily decrese (in %)? Enter ticker of the company as an answer

In [ ]:
answer_part_8 = 'UNH'

<br>

**Question 9.** What was the best month for all companies (i.e. average monthly price increase was the best)? Enter one of the following: January, February, March, April, May, June, July, August, September, October, November, December

In [ ]:
answer_part_9 = 'January'

In [189]:
def avg_monthly_change(df):
    monthly_df = df.groupby('Ticker').apply(get_open_close)
    avg_change = ((monthly_df['Close'] / monthly_df['Open'] - 1) * 100).mean()
    return pd.Series({'Change': avg_change})

month_names = ['January', 'February', 'March', 'April', 'May', 'June', 'July',
    'August', 'September', 'October', 'November', 'December'
]
month_names = {i + 1: name for i, name in enumerate(month_names)}

monthly_change = df_2019_sort.groupby('Month').apply(avg_monthly_change).reset_index()
monthly_change['MonthName'] = monthly_change['Month'].map(month_names)
monthly_change.sort_values(by=['Change'], ascending=False)

Month    Change  MonthName
0       1  7.409273    January
5       6  7.062817       June
8       9  3.443022  September
1       2  3.382967   February
10     11  2.573628   November
11     12  2.120015   December
3       4  1.984414      April
9      10  0.697505    October
2       3  0.303609      March
6       7 -0.240822       July
7       8 -2.751223     August
4       5 -6.664097        May

<br>

**Question 10.** What was the worst month for all companies (i.e. average monthly price increase was the worst)? Enter one of the following: January, February, March, April, May, June, July, August, September, October, November, December

answer_part_10 = May

In [193]:
answer_part_10 = 'May'